server/src/main/java/org/opensearch/wlm/cancellation/TaskCanceller.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.wlm.cancellation;
+
+import org.opensearch.cluster.metadata.QueryGroup;
+import org.opensearch.tasks.CancellableTask;
+import org.opensearch.tasks.TaskCancellation;
+import org.opensearch.wlm.MutableQueryGroupFragment.ResiliencyMode;
+import org.opensearch.wlm.QueryGroupLevelResourceUsageView;
+import org.opensearch.wlm.QueryGroupTask;
+import org.opensearch.wlm.ResourceType;
+import org.opensearch.wlm.WorkloadManagementSettings;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BooleanSupplier;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import static org.opensearch.wlm.tracker.QueryGroupResourceUsageTrackerService.TRACKED_RESOURCES;
+
+/**
+ * Manages the cancellation of tasks enforced by QueryGroup thresholds on resource usage criteria.
+ * This class utilizes a strategy pattern through {@link MaximumResourceTaskSelectionStrategy} to identify tasks that exceed
+ * predefined resource usage limits and are therefore eligible for cancellation.
+ *
+ * <p>The cancellation process is initiated by evaluating the resource usage of each QueryGroup against its
+ * resource limits. Tasks that contribute to exceeding these limits are selected for cancellation based on the
+ * implemented task selection strategy.</p>
+ *
+ * <p>Instances of this class are configured with a map linking QueryGroup IDs to their corresponding resource usage
+ * views, a set of active QueryGroups, and a task selection strategy. These components collectively facilitate the
+ * identification and cancellation of tasks that threaten to breach QueryGroup resource limits.</p>
+ *
+ * @see MaximumResourceTaskSelectionStrategy
+ * @see QueryGroup
+ * @see ResourceType
+ */
+public class TaskCanceller {
+    public static final double MIN_VALUE = 1e-9;
+
+    private final WorkloadManagementSettings workloadManagementSettings;
+    private final TaskSelectionStrategy taskSelectionStrategy;
+    // a map of QueryGroupId to its corresponding QueryGroupLevelResourceUsageView object
+    private final Map<String, QueryGroupLevelResourceUsageView> queryGroupLevelResourceUsageViews;
+    private final Collection<QueryGroup> activeQueryGroups;
+    private final Collection<QueryGroup> deletedQueryGroups;
+    private BooleanSupplier isNodeInDuress;
+
+    public TaskCanceller(
+        WorkloadManagementSettings workloadManagementSettings,
+        MaximumResourceTaskSelectionStrategy taskSelectionStrategy,
```

Comment:
why are we using concrete strategy class here?

Reply:
Good Catch! I will address this